samples/bluetooth/peripheral_hids_keyboard/src/main.c

```diff
@@ -0,0 +1,629 @@
+		return;
+	}
+
+	for (size_t i = 0; i < CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT; i++) {
+		if (!conn_mode[i].conn) {
+			conn_mode[i].conn = conn;
+			conn_mode[i].in_boot_mode = false;
+			return;
+		}
+	}
+}
+
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	int err;
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected from %s (reason %u)\n", addr, reason);
+
+	err = hids_notify_disconnected(&hids_obj, conn);
+
+	if (err) {
+		printk("Failed to notify HID service about disconnection\n");
+	}
+
+	for (size_t i = 0; i < CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT; i++) {
+		if (conn_mode[i].conn == conn) {
+			conn_mode[i].conn = NULL;
+			break;
+		}
+	}
+}
+
+
+static void security_changed(struct bt_conn *conn, bt_security_t level)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Security changed: %s level %u\n", addr, level);
+}
+
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed,
+};
+
+
+
+static void key_pack_queue(char *string, u8_t len)
+{
+	struct key_pack keys;
+	int err;
+
+	__ASSERT(len <= KEYS_MAX_LEN, "String is too long.");
+
+	printk("Getting key: %02X\n", (char) *string);
+
+	keys.elements = string;
+	keys.len = len;
+
+	err = k_msgq_put(&hids_queue, &keys, K_NO_WAIT);
+
+	if (err) {
+		printk("No space in the queue for button pressed\n");
+		return;
+	}
+}
+
+
+static void caps_lock_handler(const struct hids_rep *rep)
+{
+	u8_t report_val = (*rep->data) & OUTPUT_REPORT_BIT_MASK_CAPS_LOCK;
+
+	if (!caps_on && (report_val != 0)) {
+		gpio_pin_write(led_dev, CAPS_LOCK_LED, 0);
+
+		key_pack_queue(caps_on_key_scan_str,
+			       sizeof(caps_on_key_scan_str));
+		caps_on = true;
+	} else if (caps_on && (report_val == 0)) {
+		gpio_pin_write(led_dev, CAPS_LOCK_LED, 1);
+
+		key_pack_queue(caps_off_key_scan_str,
+			       sizeof(caps_off_key_scan_str));
+		caps_on = false;
+	}
+}
+
+
+static void hids_outp_rep_handler(const struct hids_rep *rep,
+				  struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+	printk("Output report has been received %s\n", addr);
+	caps_lock_handler(rep);
+}
+
+
+static void hids_boot_kb_outp_rep_handler(const struct hids_rep *rep,
+					  struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+	printk("Boot Keyboard Output report has been received %s\n", addr);
+	caps_lock_handler(rep);
+}
+
+
+static void hids_pm_evt_handler(enum hids_pm_evt evt, struct bt_conn *conn)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	size_t i;
+
+	for (i = 0; i < CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT; i++) {
+		if (conn_mode[i].conn == conn) {
+			break;
+		}
+	}
+
+	if (i >= CONFIG_BT_GATT_HIDS_MAX_CLIENT_COUNT) {
+		return;
```

Comment:
Could it be useful to print something in this case? I wonder if this couldn't be an `assert`.

Reply:
I would print here something. It feels not good to add an assert here as this is theoretically possible that device would be disconnected before pm is processed (I am not sure about it). It is to much externaly dependent.